### PR TITLE
Retry following redirects on 400 (not 404), 502, 503, & 504 up to 5 times

### LINKED ISSUE #761: retry redirect on 502, 503, 504
```
2021-08-31T15:15:42.2110260Z _________________ test_download_000027_resume[draft-<lambda>0] _________________
2021-08-31T15:15:42.2111008Z dandi/tests/test_download.py:108: in test_download_000027_resume
2021-08-31T15:15:42.2111761Z     download(url, tmp_path, get_metadata=False)
2021-08-31T15:15:42.2112393Z dandi/download.py:59: in download
2021-08-31T15:15:42.2112961Z     parsed_url = parse_dandi_url(urls[0])
2021-08-31T15:15:42.2113569Z dandi/dandiarchive.py:421: in parse
2021-08-31T15:15:42.2114186Z     new_url = cls.follow_redirect(url)
2021-08-31T15:15:42.2114830Z dandi/dandiarchive.py:538: in follow_redirect
2021-08-31T15:15:42.2115616Z     f"Response for getting {url} to redirect returned {r.status_code}."
2021-08-31T15:15:42.2117463Z E   dandi.exceptions.FailedToConnectError: Response for getting https://dandiarchive.org/dandiset/000027/draft to redirect returned 504. Please verify that it is a URL related to dandiarchive and supported by dandi client
```
from tests runs of https://github.com/dandi/dandi-cli/pull/760#pullrequestreview-742868211 . Let's add those other error codes into the list (has 404 only ATM) and extend retrying to e.g. 5 trials.

### PR DESCRIPTION
**Pull request by jwodder**
Closes #761.

### CODE REVIEW

File: dandi/dandiarchive.py [hunk 1]
```diff
@@ -526,8 +526,8 @@ def follow_redirect(url):
         i = 0
         while True:
             r = requests.head(url, allow_redirects=True)
-            if r.status_code == 404:
-                if i < 3:
+            if r.status_code in (404, 502, 503, 504):
```
Comment (yarikoptic) [inline dandi/dandiarchive.py:529]:
```suggestion
            if r.status_code in (400, 502, 503, 504):
```
per brief investigation in https://github.com/dandi/dandi-cli/issues/261#issuecomment-909631735

**Comment by yarikoptic** [comment]
I guess a unittest needs fix ups to reflect 404 -> 400 change
```
________________________ test_parse_dandi_url_not_found ________________________
dandi/tests/test_dandiarchive.py:240: in test_parse_dandi_url_not_found
    parse_dandi_url("https://dandiarchive.org/dandiset/999999")
dandi/dandiarchive.py:421: in parse
    new_url = cls.follow_redirect(url)
dandi/dandiarchive.py:537: in follow_redirect
    raise FailedToConnectError(
E   dandi.exceptions.FailedToConnectError: Response for getting https://dandiarchive.org/dandiset/999999 to redirect returned 404. Please verify that it is a URL related to dandiarchive and supported by dandi client
```

**Comment by jwodder** [comment]
@yarikoptic Fixed.

**Comment by yarikoptic** [comment]
Great, thank you!

**Pull request merged**